[lldb] Introduce ScriptedFrameProvider for real threads

medismailben opened the pull request:
This patch extends ScriptedFrame to work with real (non-scripted) threads,
enabling frame providers to synthesize frames for native processes.

Previously, ScriptedFrame only worked within ScriptedProcess/ScriptedThread
contexts. This patch decouples ScriptedFrame from ScriptedThread, allowing
users to augment or replace stack frames in real debugging sessions for use
cases like custom calling conventions, reconstructing corrupted frames from
core files, or adding diagnostic frames.

Key changes:

- ScriptedFrame::Create() now accepts ThreadSP instead of requiring
ScriptedThread, extracting architecture from the target triple rather
than ScriptedProcess.arch

- Added SBTarget::RegisterScriptedFrameProvider() and
ClearScriptedFrameProvider() APIs, with Target storing a
SyntheticFrameProviderDescriptor template for new threads

- Added "target frame-provider register/clear" commands for CLI access

- Thread class gains LoadScriptedFrameProvider(), ClearScriptedFrameProvider(),
and GetFrameProvider() methods for per-thread frame provider management

- New SyntheticStackFrameList overrides FetchFramesUpTo() to lazily provide
frames from either the frame provider or the real stack

This enables practical use of the SyntheticFrameProvider infrastructure in
real debugging workflows.

rdar://161834688

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>

vogelsgesang reviewed (commented):
Being able to modify the frame list from a script would indeed be amazing - thanks for looking into this! 🙂 

As it happens, I am currently looking into gdb's "frame filter" API, which serves a very similar use case - and I am wondering if we should take some inspiration from the gdb APIs here. Gdb's frame filter API differs from the API in this PR in the following points:

* **simpler API**: gdb's API only has a single function `filter(frame_iter) -> frame_iter` function. (This PR currently introduces two functions: `get_merge_strategy` and `get_stack_frames`)
* **lazy stack unwinding**: afaict, your patch leads to eager materialization of all stack frames. Given that stacks can be pretty deep, it would be preferable to only create those stack frames lazily, as the users inspects the stack. gdb achieves this by only advancing the returned iterator as needed
* **more flexible merging strategies**: in gdb, I can use all of Python's iterator support (generator expressions, `yield`, ...)
* **global registration**: in gdb, frame filters are registered globally for all threads. In this PR, the frame providers are registered for each thread individually
* **multiple frame filters**: gdb supports registering multiple frame filters at the same time. I didn't use that mechanism, yet, but afaict they are simply chained

The relevant pieces of gdb's documentation:
* [Writing a frame filter](https://www.sourceware.org/gdb/current/onlinedocs/gdb.html/Writing-a-Frame-Filter.html) - high-level introduction
* [Frame filter API](https://www.sourceware.org/gdb/current/onlinedocs/gdb.html/Frame-Filter-API.html) - corresponding API reference
* [Frame decorators API](https://www.sourceware.org/gdb/current/onlinedocs/gdb.html/Frame-Decorator-API.html#Frame-Decorator-API) - close equivalent to lldb's `ScriptedFrame`
* [Frame filter management](https://www.sourceware.org/gdb/current/onlinedocs/gdb.html/Frame-Filter-Management.html#Frame-Filter-Management) - end users can change enable, disable and reorder frame filters

My motivation for using frame filters in gdb are C++ couroutines. I want to add frames for asynchronous operations. For more background see LLVM's documentatin on [Async Stack Traces](https://clang.llvm.org/docs/DebuggingCoroutines.html#async-stack-traces), in particular the `coro bt` example. At least for my use case, gdb's design decisions (global registration, flexible merging strategy, ...) are a pretty good fit. But not sure, maybe your current design is a better fit for your particular use case? Which use case are you envisioning?

By the way: would discourse be better-suited for discussing the best approach here? Happy to copy my reply over to discourse, if you would start a thread

medismailben commented on lldb/source/Target/Thread.cpp:
```diff
@@ -1439,13 +1444,133 @@ void Thread::CalculateExecutionContext(ExecutionContext &exe_ctx) {
+
+        // Add real frames after scripted frames (shifted indices)
+        for (uint32_t i = 0; i < num_real; i++) {
+          StackFrameSP real_frame = real_frames_sp->GetFrameAtIndex(i);
```
Only if the user picks the `eScriptedFrameProviderMergeStrategyPrepend` or  `eScriptedFrameProviderMergeStrategyAppend ` merge strategies which is necessary.
You probably picked that up by looking at the other files that the default option is `eScriptedFrameProviderMergeStrategyReplace`

JDevlieghere commented:
> A non-goal for this PR and for lldb in general is copy what gdb or other debuggers do.

While compatibility is not generally a goal, there's also no reason to diverge for the sake of it. However, what I think Adrian is suggesting here, is that we can benefit from building on top of the learnings of an existing implementation. I think he raises some good points that are worth evaluating in the context of LLDB.

vogelsgesang commented:
>> A non-goal for this PR and for lldb in general is copy what gdb or other debuggers do.
> While compatibility is not generally a goal, there's also no reason to diverge for the sake of it. However, what I think Adrian is suggesting here, is that we can benefit from building on top of the learnings of an existing implementation.

yes. Thanks for summing it up!

### Use cases

> My (Adrian's) motivation for using frame filters in gdb are C++ coroutines.

I polished my existing script, so I could post it as PR #162145. The most interesting piece is `CppCoroutineFrameFilter`. With this filter, I get the backtrace

```
(gdb) bt
#0  write_output(std::basic_string_view<char, std::char_traits<char> >) [clone .resume] (contents=...) at async-task-example.cpp:6
[async] greet () at async-task-example.cpp:12
[async] std::__n4861::coroutine_handle<std::__n4861::noop_coroutine_promise>::__frame::__dummy_resume_destroy() at <...>
#1  0x000055555555a0f8 in std::__n4861::coroutine_handle<task::promise_type>::resume (this=0x7fffffffd5a8) at /usr/include/c++/14/coroutine:242
#2  0x0000555555557bb9 in task::syncStart (this=0x7fffffffd5a8) at async-task-library.hpp:78
#3  0x00005555555552de in main () at async-task-example.cpp:18
```

The two lines prefixed with `[async]` are the additional frames injected by the `yield from _create_coroutine_frames(parent_coro, inferior_frame)` in the frame filter. ([Source code](https://clang.llvm.org/docs/DebuggingCoroutines.html#stack-traces-of-in-flight-coroutines), in case anyone wants to reproduce this example)

> My (@medismailben's) current use case for this is to translate CPython frames into something that's more understandable, to hopefully improve debugging python / C++ interoperability in lldb.

Could you provide an example for a `ScriptedFrameProvider` implementation pretty-printing the CPython frames? Or at least the intended `bt` output which you are envisioning for an example Python program?

I am currently slightly struggling with imagining the intended usage of the proposed APIs, and such an example might help me. (Of course only if possible and not blocked, e.g., due to intellectual property issues)

vogelsgesang commented on lldb/source/Target/Thread.cpp:
```diff
@@ -1439,13 +1444,133 @@ void Thread::CalculateExecutionContext(ExecutionContext &exe_ctx) {
+      uint32_t num_real = real_frames_sp->GetNumFrames(true);
+      uint32_t num_scripted = scripted_frames_sp->GetNumFrames(false);
+
+      switch (strategy) {
```
> main source of truth should be the Thread itself

Not sure I can follow. Can you elaborate?

medismailben commented on lldb/examples/python/templates/scripted_frame_provider.py:
```diff
@@ -0,0 +1,147 @@
+
+        # Attach a frame provider to a thread
+        thread = process.GetSelectedThread()
+        error = lldb.SBError()
```
Good catch, it's actually the return value of `SBThread::SetScriptedFrameProvider`. I'll update the doc.

JDevlieghere reviewed (commented):
I quickly looked at the private parts of the change. I had a slightly different design in mind, which I pitched to Ismail in person this morning. I created a quick sequence diagram to show it in action. 

<img width="2164" height="3536" alt="Image" src="https://github.com/user-attachments/assets/44027111-659e-45af-be41-c81c819d6003" />

Here, we start by asking for the synthetic frame at index 3. The synthetic stack frame list realizes it has no frames, so it queries the stack frame provider. The latter adds 3 frames to the synthetic stack frame list. When next we get a call for frame 2, we don't have to query the provider, since we already have that stack frame. However, when asking for the frame at index 4, we have to query the provider again, and it populates the synthetic stack frame list with a fourth frame.

jimingham commented:
A couple of comments on Jonas' comments.  First, I don't think it's necessary to have a privileged "real stack frame list" and the a "scripted stack frame provider" that refers to it.  There's no reason this can't be a chain of providers.  The important thing is that each provider is given the stackframelist it uses as its source material that it asks when it needs more frames to work on.  It shouldn't need to know where those stack frames are coming from.

Also, I agree the StackFrameList providers should be held by the target because among other things users need to set them so they are available the first time a new process stops, which is straightforward if the target holds them.  OTOH, I do think that there should be some "AppropriateForThread" API in the provider.  If you know that all the threads you are going to work on have a particular thread name or belong to a specific queue or whatever, there should be a way to express that so the we don't end up asking the providers to do a bunch of useless work.

jimingham commented:
People have expressed an interest in being able to write these stack frame list providers in C++ (both internally and through the SB API) as well as having the capability exposed to the scripting language.  

I think that's a reasonable request, and also we should start supporting this more generally.  It's kind of bogus, for instance, that if you are in lldb_private you can write data formatters in C++, but you can't do that through the SB API's, you have to use Python for that.  We've thought about this for a while (there are 4 cryptic references to lldb.SBSyntheticValueProvider in lldb that are intended for that purpose but not implemented).  So I'd like to keep that in mind as we decide how we're going to offer these capabilities.

Doing that is what I meant "will be much easier to do with a GetFrameAtIndex API than a C++ API that interacts with a generator."

I also am guessing that it would be easier to wrap that API in whatever generator is natural to the scripting language you are forwarding these API's to.

jimingham commented on lldb/bindings/interface/SBFrameListExtensions.i:
```diff
@@ -0,0 +1,46 @@
+            return int(self.GetSize())
+
+        def __getitem__(self, key):
+            count = len(self)
```
Why do you need to count the stack frames in `__getitem__`?  That has the side-effect of forcing the generation of all the stack frames in the list, which we should only do when directly asked (like when asked for the number of stack frames).  
But also, I don't see how it helps.  If you ran GetFrameAtIndex on an index already fetched, it will return that.  If you run GetFrameAtIndex on a number between what was fetched and the number of frames in the stack, it will realize frames up to that number and stop.  If you ask for more frames than exist, then it will fetch all the frames available, find that your number is too big and return None.
So I don't think you get anything out of forcing the count here.

jimingham commented on lldb/include/lldb/API/SBFrameList.h:
```diff
@@ -0,0 +1,55 @@
+private:
+  SBFrameList(const lldb::StackFrameListSP &frame_list_sp);
+
+  void SetOpaque(const lldb::StackFrameListSP &frame_list_sp);
```
I'd call this SetFrameList, the fact that the ivar is called "m_opaque_sp" is too much of an implementation detail and SetOpaque sounds like it's making the frame list opaque somehow.

jimingham commented on lldb/include/lldb/Target/Thread.h:
```diff
@@ -1336,8 +1338,18 @@ class Thread : public std::enable_shared_from_this<Thread>,
     return StructuredData::ObjectSP();
   }
 
+<<<<<<< HEAD
```
Would this go away if I wasn't looking at the patches one by one?

medismailben replied:
Yes

jimingham commented on lldb/test/API/python_api/frame_list/TestSBFrameList.py:
```diff
@@ -0,0 +1,172 @@
+
+    def frame_list_api(self):
+        """Test SBThread.GetFrames() returns a valid SBFrameList."""
+        exe = self.getBuildArtifact("a.out")
```
This section is what lldbutil.run_to_source_breakpoint does, but your version does a little less sanity checking.  Better to use run_to_source_breakpoint.

jimingham commented on lldb/source/Interpreter/ScriptInterpreter.cpp:
```diff
@@ -150,6 +150,11 @@ ScriptInterpreter::GetOpaqueTypeFromSBExecutionContext(
   return exe_ctx.m_exe_ctx_sp;
 }
 
+lldb::StackFrameListSP ScriptInterpreter::GetOpaqueTypeFromSBFrameList(
```
OpaqueType is a confusing name for the underlying StackFrameList.  GetStackFrameListFromSBFrameList seems much clearer.  Also, why isn't this a method on SBFrameList?

medismailben replied:
1. This is how we've been doing it. Renaming these methods would be orthogonal to this PR.
2. We can't have it be in SBFrameList because each ScriptingInterface will have to talk to SBAPI classes directly would be a layering violation. We had to use the ScriptInterpreter for that.

jimingham commented on lldb/include/lldb/Target/SyntheticFrameProvider.h:
```diff
@@ -0,0 +1,115 @@
+
+  /// Optional list of thread IDs to which this provider applies.
+  /// If empty, the provider applies to all threads.
+  std::vector<lldb::tid_t> thread_ids;
```
You should allow thread names here as well.  Thread ID's aren't stable and so aren't going to be easy to use in general, but thread names can be known before running the binary and so are more useful for denoting "interesting threads".

jimingham commented on lldb/include/lldb/Target/SyntheticFrameProvider.h:
```diff
@@ -0,0 +1,115 @@
+    // If no thread IDs specified, applies to all threads
+    if (thread_ids.empty())
+      return true;
+
```
For instance, here's where you would want to find the Thread for `tid` and get its name to compare to the name list you really ought to have.

jimingham commented on lldb/include/lldb/Target/SyntheticFrameProvider.h:
```diff
@@ -0,0 +1,115 @@
+///
+/// This struct contains the metadata needed to instantiate a frame provider
+/// and optional filters to control which threads it applies to.
+struct SyntheticFrameProviderDescriptor {
```
I'll have to read the next patch to see how this is used...

medismailben replied:
I've changed SyntheticFrameProviderDescriptor to hold a vector of `ThreadSpec` and `AppliesToThread` to take a `const Thread&` instead of the `tid_t`

jimingham commented on lldb/source/Plugins/Process/scripted/ScriptedFrame.h:
```diff
@@ -33,8 +31,28 @@ class ScriptedFrame : public lldb_private::StackFrame {
+  /// \param[in] args_sp
+  ///     Arguments to pass to the frame creation.
+  ///
+  /// \param[in] script_object
```
This parameter is also optional, it would be good to say here what it means to have and not have it.

jimingham commented on lldb/source/Plugins/SyntheticFrameProvider/ScriptedFrameProvider/ScriptedFrameProvider.cpp:
```diff
@@ -0,0 +1,232 @@
+  if (!m_interface_sp) {
+    error = Status::FromErrorString(
+        "cannot create scripted frame provider: Script interpreter couldn't "
+        "create Scripted Frame Provider Interface");
```
This is not helpful...

jimingham commented on lldb/source/Plugins/SyntheticFrameProvider/ScriptedFrameProvider/ScriptedFrameProvider.cpp:
```diff
@@ -0,0 +1,232 @@
+  }
+
+  error.Clear();
+}
```
Since this constructor can fail, you need something to say it didn't succeed.  Looks like you are using !m_interface_sp, but if you get an error at this stage, then m_interface_sp has a value, but there was still a construction error?

jimingham commented on lldb/source/Plugins/SyntheticFrameProvider/ScriptedFrameProvider/ScriptedFrameProvider.cpp:
```diff
@@ -0,0 +1,232 @@
+    return llvm::createStringError(
+        "cannot get stack frame: Scripted frame provider not initialized");
+
+  auto create_frame_from_dict =
```
Do you gain anything by having these be closures rather than methods?  They make the function harder to read because you have to peer past these definitions first, and so far as I can tell, they only capture "this" and so are really just methods.

jimingham commented on lldb/source/Plugins/SyntheticFrameProvider/ScriptedFrameProvider/ScriptedFrameProvider.h:
```diff
@@ -0,0 +1,62 @@
+    return "ScriptedFrameProvider";
+  }
+
+  static llvm::Expected<lldb::SyntheticFrameProviderSP>
```
At some point we're going to want to chain these providers together.  To do that, the ScriptedFrameProvider needs to take a StackFrameList, not a thread, since it might be acting on a StackFrameList produced by something in the chain of StackFrameProviders.  So it's probably best to start with that.  StackFrameList's know their thread (though there isn't an accessor for it that's trivial to add.)  So you wouldn't be losing information.

medismailben replied:
+1

medismailben commented on lldb/source/Plugins/SyntheticFrameProvider/ScriptedFrameProvider/ScriptedFrameProvider.h:
```diff
@@ -0,0 +1,62 @@
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLDB_PLUGINS_SYNTHETICFRAMEPROVIDER_SCRIPTEDFRAMEPROVIDER_SCRIPTEDFRAMEPROVIDER_H
```
That's how it should be done according to the coding standards: https://llvm.org/docs/CodingStandards.html#header-guard

medismailben commented on lldb/source/Plugins/SyntheticFrameProvider/ScriptedFrameProvider/ScriptedFrameProvider.cpp:
```diff
@@ -0,0 +1,232 @@
+    return;
+  }
+
+  StructuredData::ObjectSP object_sp = *obj_or_err;
```
The scripted interface (`m_interface_sp`) manages it.

medismailben commented on lldb/source/API/SBTarget.cpp:
```diff
@@ -2408,3 +2410,55 @@ lldb::SBMutex SBTarget::GetAPIMutex() const {
+  }
+
+  // Extract the dictionary from SBStructuredData
+  StructuredData::DictionarySP dict_sp;
```
I've addressed the first comment but in order the second one if a bit tricky ... I'll investigate this in a follow-up.

medismailben commented on lldb/source/Target/Thread.cpp:
```diff
@@ -1439,13 +1445,59 @@ void Thread::CalculateExecutionContext(ExecutionContext &exe_ctx) {
+        if (descriptor && descriptor->IsValid()) {
+          // Check if this descriptor applies to this thread
+          if (descriptor->AppliesToThread(GetID())) {
+            if (llvm::Error error = LoadScriptedFrameProvider()) {
```
I was also unhappy with this so I re-wrote it, let me know if you prefer it better like this.

medismailben commented:
In order to simplify reviewing this, I've split the commits in different PRs: #166651, #166664, #166662 and this one.

medismailben commented:
I've landed all other other PRs so we can just focus on the core functionality of synthetic / scripted frame providers.

JDevlieghere commented on lldb/include/lldb/API/SBTarget.h:
```diff
@@ -986,6 +986,34 @@ class LLDB_API SBTarget {
+  ///     An error object indicating success or failure.
+  lldb::SBError RemoveScriptedFrameProvider(const char *provider_name);
+
+  /// Clear the scripted frame provider for this target.
```
How many scripted frame providers can exist per target? If it's only one, why do we need both `RemoveScriptedFrameProvider` and `ClearScriptedFrameProvider`. If it's multiple, then this method should be plural.

medismailben replied:
See https://github.com/llvm/llvm-project/pull/161870#discussion_r2482534752

Currently the target can have multiple scripted frame provider descriptor however the threads can only have 1 provider at a time. I'm planning on being able to chain providers (both internal plugins & scripted implementations) in a follow-up. I actually think this API should be called `RemoveSyntheticFrameProvider` not `Scripted` because the user might want to disable an lldb internal frame provider.

JDevlieghere commented on lldb/include/lldb/Target/StackFrameList.h:
```diff
@@ -211,18 +212,20 @@ class StackFrameList {
 
+  /// These two Fetch frames APIs and SynthesizeTailCallFrames are called in
+  /// GetFramesUpTo, they are the ones that actually add frames.  They must be
+  /// called with the writer end of the list mutex held.
```
Use a Doxygen group (https://www.doxygen.nl/manual/grouping.html)

```
/// @{
[...]
/// @}
```

JDevlieghere commented on lldb/include/lldb/Target/Target.h:
```diff
@@ -697,6 +698,39 @@ class Target : public std::enable_shared_from_this<Target>,
   Status Attach(ProcessAttachInfo &attach_info,
                 Stream *stream); // Optional stream to receive first stop info
 
+  // Frame provider methods
```
Remove or use Doxygen group.

JDevlieghere commented on lldb/source/Plugins/Process/scripted/ScriptedFrame.cpp:
```diff
@@ -1,4 +1,4 @@
-//===----------------------------------------------------------------------===//
+//===-- ScriptedFrame.cpp -------------------------------------------------===//
```
```suggestion
//===----------------------------------------------------------------------===//
```

JDevlieghere commented on lldb/source/Plugins/Process/scripted/ScriptedFrame.cpp:
```diff
@@ -19,50 +33,64 @@ void ScriptedFrame::CheckInterpreterAndScriptObject() const {
+  auto scripted_frame_interface = script_interp->CreateScriptedFrameInterface();
   if (!scripted_frame_interface)
-    return llvm::createStringError("failed to create scripted frame interface");
+    return llvm::createStringError("Failed to create scripted frame interface");
```
?
```suggestion
    return llvm::createStringError("failed to create scripted frame interface");
```

JDevlieghere reviewed (commented):
Please do your own pass over this PR for things that don't match the https://llvm.org/docs/CodingStandards.html. There's various instances of missing periods, incorrect capitalization, etc. Most of which I've already called out in the PRs for the dependencies.

JDevlieghere commented on lldb/include/lldb/API/SBTarget.h:
```diff
@@ -986,6 +987,42 @@ class LLDB_API SBTarget {
+  ///
+  /// \return
+  ///     An error object indicating success or failure.
+  lldb::SBError ClearScriptedFrameProvider();
```
Should this then also be `ClearSyntheticFrameProvider(s)`?

medismailben replied:
As mentioned here  https://github.com/llvm/llvm-project/pull/161870#discussion_r2505912411, the target only hold `ScriptedFrameProviderDescriptor` internally so I think calling Scripted is what makes the most sense.

JDevlieghere commented on lldb/source/Plugins/SyntheticFrameProvider/ScriptedFrameProvider/ScriptedFrameProvider.cpp:
```diff
@@ -0,0 +1,229 @@
+  StructuredData::ObjectSP object_sp = *obj_or_err;
+  if (!object_sp || !object_sp->IsValid()) {
+    error = Status::FromErrorString(
+        "cannot create scripted frame provider: failed to create valid script "
```
Similar question, what's a "script object"? Is this the ScriptedFrame object? If so, why not be more specific?

medismailben replied:
That's the scripted frame provider object. I didn't mention it because it felt repetitive to mention it given we're already saying "cannot create scripted frame provider"

JDevlieghere commented on lldb/source/Plugins/SyntheticFrameProvider/ScriptedFrameProvider/ScriptedFrameProvider.cpp:
```diff
@@ -0,0 +1,229 @@
+  if (StructuredData::UnsignedInteger *int_obj =
+          obj_sp->GetAsUnsignedInteger()) {
+    uint32_t real_frame_index = int_obj->GetValue();
+    if (real_frame_index < m_input_frames->GetNumFrames()) {
```
No braces

JDevlieghere commented on lldb/source/Target/Target.cpp:
```diff
@@ -3718,6 +3713,73 @@ Status Target::Attach(ProcessAttachInfo &attach_info, Stream *stream) {
+
+  // Clear frame providers on existing threads so they reload with new config
+  if (ProcessSP process_sp = GetProcessSP()) {
+    for (ThreadSP thread_sp : process_sp->Threads()) {
```
No braces

JDevlieghere reviewed (commented):
There are still a lot of stylistic issues left, especially w.r.t. comments.

medismailben commented on lldb/source/Plugins/Process/scripted/ScriptedFrame.h:
```diff
@@ -33,8 +31,28 @@ class ScriptedFrame : public lldb_private::StackFrame {
 
   ~ScriptedFrame() override;
 
+  /// Create a ScriptedFrame from a script object.
```
A object created in the script interpreter (in python).

jimingham commented on lldb/include/lldb/API/SBTarget.h:
```diff
@@ -986,6 +987,42 @@ class LLDB_API SBTarget {
+  /// \param[in] args_dict
+  ///     A dictionary of arguments to pass to the frame provider class.
+  ///
+  /// \param[in] thread_filter
```
How do you handle the case where when you register the provider some or all of the threads you want it to match against haven't been created yet?  I was expecting this to be algorithmic.

medismailben replied:
How can you know the thread name of id or a thread that haven't yet been created ? Also if we wanted to go that route, we would have to introduce a new SBAPI class.

medismailben replied:
That's an option but it's not very optimal since we'll be creating python instances of the frame provider just to ask if we need to associate it with said thread and discard it otherwise. Seems a bit overkill.

jimingham replied:
You don't need to create an instance of a python class to call a static method, do you?

medismailben replied:
Ok, after some offline chatting, we came to the conclusion that the static method would be a good way to let the user take care of the filtering themselves but it might still be useful to be able to do it from the command line. We should just remove it in the SBAPI.

jimingham commented on lldb/include/lldb/API/SBTarget.h:
```diff
@@ -986,6 +987,42 @@ class LLDB_API SBTarget {
 
+  /// Register a scripted frame provider for this target.
+  /// If a scripted frame provider with the same name is already registered on
+  /// this target, it will be overwritten.
```
I think it would make more sense if this API returned a token for the registered provider, and that would be the way you referred to it.  That would make supporting this more general usage straightforward.

medismailben replied:
> I can imagine wanting to use the same class, and parametrizing different uses instead by passing different content in the args_dict. Why is this impossible?

Can you elaborate on how you'd use that ? Do you applying the same scripted frame provider multiple times but with different thread filters for instance ?

JDevlieghere commented on lldb/source/Target/Target.cpp:
```diff
@@ -3718,6 +3713,69 @@ Status Target::Attach(ProcessAttachInfo &attach_info, Stream *stream) {
+      m_frame_provider_descriptors_mutex);
+  m_frame_provider_descriptors[name] = descriptor;
+
+  // Clear frame providers on existing threads so they reload with new config
```
```suggestion
  // Clear frame providers on existing threads so they reload with new config.
```

jimingham commented on lldb/include/lldb/Target/Target.h:
```diff
@@ -697,6 +698,37 @@ class Target : public std::enable_shared_from_this<Target>,
                 Stream *stream); // Optional stream to receive first stop info
 
+  /// Add or update a scripted frame provider descriptor for this target.
+  /// The descriptor's name is extracted from its scripted metadata class name.
```
This should be a token returned from AddScriptedFrameProviderDescriptor, as we discussed above.

JDevlieghere commented on lldb/include/lldb/API/SBTarget.h:
```diff
@@ -986,6 +987,42 @@ class LLDB_API SBTarget {
+  ///
+  /// \return
+  ///     An error object indicating success or failure.
+  lldb::SBError RemoveScriptedFrameProvider(const char *provider_name);
```
Looks like this is untested.

JDevlieghere reviewed (commented):
LGTM if Jim's happy.

jimingham commented on lldb/source/Commands/CommandObjectTarget.cpp:
```diff
@@ -5392,6 +5393,191 @@ class CommandObjectTargetDump : public CommandObjectMultiword {
+            "Remove a registered frame provider from the target by name.",
+            "target frame-provider remove <provider-name>",
+            eCommandRequiresTarget) {
+    AddSimpleArgumentList(eArgTypeClassName, eArgRepeatPlain);
```
DOn't forget to change the argument type when you convert these to a token.

jimingham commented on lldb/source/Commands/CommandObjectTarget.cpp:
```diff
@@ -5392,6 +5393,191 @@ class CommandObjectTargetDump : public CommandObjectMultiword {
+
+protected:
+  void DoExecute(Args &command, CommandReturnObject &result) override {
+    Target *target = m_exe_ctx.GetTargetPtr();
```
Should these be added to the dummy target if there isn'a an actual target yet?

medismailben commented on lldb/include/lldb/Target/Target.h:
```diff
@@ -697,6 +698,37 @@ class Target : public std::enable_shared_from_this<Target>,
+  void ClearScriptedFrameProviderDescriptors();
+
+  /// Get all scripted frame provider descriptors for this target.
+  const llvm::StringMap<SyntheticFrameProviderDescriptor> &
```
This is passing a const reference so it's fine.

jimingham commented on lldb/source/Commands/CommandObjectTarget.cpp:
```diff
@@ -5401,6 +5402,199 @@ class CommandObjectTargetDump : public CommandObjectMultiword {
+  CommandObjectTargetFrameProviderRemove(CommandInterpreter &interpreter)
+      : CommandObjectParsed(
+            interpreter, "target frame-provider remove",
+            "Remove a registered frame provider from the target by name.",
```
"by name" -> "by id" and "<provider-name>" -> "<provider-id>" just below.

jimingham reviewed (approved):
LGTM with the trivial provider -> providers swap.

medismailben committed:
[lldb] Add support for ScriptedFrame with real threads

This patch extends ScriptedFrame to work with real (non-scripted) threads,
enabling frame providers to synthesize frames for native processes.

Previously, ScriptedFrame only worked within ScriptedProcess/ScriptedThread
contexts. This patch decouples ScriptedFrame from ScriptedThread, allowing
users to augment or replace stack frames in real debugging sessions for use
cases like custom calling conventions, reconstructing corrupted frames from
core files, or adding diagnostic frames.

Key changes:

- ScriptedFrame::Create() now accepts ThreadSP instead of requiring
ScriptedThread, extracting architecture from the target triple rather
than ScriptedProcess.arch

- Added SBTarget::RegisterScriptedFrameProvider() and
ClearScriptedFrameProvider() APIs, with Target storing a
SyntheticFrameProviderDescriptor template for new threads

- Added "target frame-provider register/clear" commands for CLI access

- Thread class gains LoadScriptedFrameProvider(), ClearScriptedFrameProvider(),
and GetFrameProvider() methods for per-thread frame provider management

- New SyntheticStackFrameList overrides FetchFramesUpTo() to lazily provide
frames from either the frame provider or the real stack

This enables practical use of the SyntheticFrameProvider infrastructure in
real debugging workflows.

rdar://161834688

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>

Michael137 commented:
Looks like this is failing on the public macOS bots (both x86 and arm64): https://green.lab.llvm.org/job/llvm.org/view/LLDB/job/as-lldb-cmake/36221/execution/node/106/log/
```
09:27:59  ======================================================================
09:27:59  FAIL: test_append_frames (TestScriptedFrameProvider.ScriptedFrameProviderTestCase)
09:27:59     Test that we can add frames after real stack.
09:27:59  ----------------------------------------------------------------------
09:27:59  Traceback (most recent call last):
09:27:59    File "/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/llvm-project/lldb/test/API/functionalities/scripted_frame_provider/TestScriptedFrameProvider.py", line 122, in test_append_frames
09:27:59      self.assertEqual(new_frame_count, original_frame_count + 1)
09:27:59  AssertionError: 5 != 6
09:27:59  Config=arm64-/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/lldb-build/bin/clang
09:27:59  ======================================================================
09:27:59  FAIL: test_applies_to_thread (TestScriptedFrameProvider.ScriptedFrameProviderTestCase)
09:27:59     Test that applies_to_thread filters which threads get the provider.
09:27:59  ----------------------------------------------------------------------
09:27:59  Traceback (most recent call last):
09:27:59    File "/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/llvm-project/lldb/test/API/functionalities/scripted_frame_provider/TestScriptedFrameProvider.py", line 218, in test_applies_to_thread
09:27:59      self.assertEqual(
09:27:59  AssertionError: 5 != 1 : Thread with ID 1 should have 1 synthetic frame
09:27:59  Config=arm64-/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/lldb-build/bin/clang
09:27:59  ======================================================================
09:27:59  FAIL: test_prepend_frames (TestScriptedFrameProvider.ScriptedFrameProviderTestCase)
09:27:59     Test that we can add frames before real stack.
09:27:59  ----------------------------------------------------------------------
09:27:59  Traceback (most recent call last):
09:27:59    File "/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/llvm-project/lldb/test/API/functionalities/scripted_frame_provider/TestScriptedFrameProvider.py", line 84, in test_prepend_frames
09:27:59      self.assertEqual(new_frame_count, original_frame_count + 2)
09:27:59  AssertionError: 5 != 7
09:27:59  Config=arm64-/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/lldb-build/bin/clang
09:27:59  ======================================================================
09:27:59  FAIL: test_remove_frame_provider_by_id (TestScriptedFrameProvider.ScriptedFrameProviderTestCase)
09:27:59     Test that RemoveScriptedFrameProvider removes a specific provider by ID.
09:27:59  ----------------------------------------------------------------------
09:27:59  Traceback (most recent call last):
09:27:59    File "/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/llvm-project/lldb/test/API/functionalities/scripted_frame_provider/TestScriptedFrameProvider.py", line 272, in test_remove_frame_provider_by_id
09:27:59      self.assertEqual(thread.GetNumFrames(), 3, "Should have 3 synthetic frames")
09:27:59  AssertionError: 5 != 3 : Should have 3 synthetic frames
09:27:59  Config=arm64-/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/lldb-build/bin/clang
09:27:59  ======================================================================
09:27:59  FAIL: test_replace_all_frames (TestScriptedFrameProvider.ScriptedFrameProviderTestCase)
09:27:59     Test that we can replace the entire stack.
09:27:59  ----------------------------------------------------------------------
09:27:59  Traceback (most recent call last):
09:27:59    File "/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/llvm-project/lldb/test/API/functionalities/scripted_frame_provider/TestScriptedFrameProvider.py", line 41, in test_replace_all_frames
09:27:59      self.assertEqual(thread.GetNumFrames(), 3, "Should have 3 synthetic frames")
09:27:59  AssertionError: 5 != 3 : Should have 3 synthetic frames
09:27:59  Config=arm64-/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/lldb-build/bin/clang
09:27:59  ======================================================================
09:27:59  FAIL: test_scripted_frame_objects (TestScriptedFrameProvider.ScriptedFrameProviderTestCase)
09:27:59     Test that provider can return ScriptedFrame objects.
09:27:59  ----------------------------------------------------------------------
09:27:59  Traceback (most recent call last):
09:27:59    File "/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/llvm-project/lldb/test/API/functionalities/scripted_frame_provider/TestScriptedFrameProvider.py", line 159, in test_scripted_frame_objects
09:27:59      self.assertEqual(frame0.GetFunctionName(), "custom_scripted_frame_0")
09:27:59  AssertionError: 'thread_func(int)' != 'custom_scripted_frame_0'
09:27:59  - thread_func(int)
09:27:59  + custom_scripted_frame_0
09:27:59  
09:27:59  Config=arm64-/Users/ec2-user/jenkins/workspace/llvm.org/as-lldb-cmake/lldb-build/bin/clang
09:27:59  ----------------------------------------------------------------------
09:27:59  Ran 6 tests in 14.242s
09:27:59  
09:27:59  FAILED (failures=6)
```

Reverting for now

DavidSpickett commented:
I'm looking at the Arm failures. First part is probably a Thumb function, second part might be a codegen difference.

DavidSpickett commented:
The PC value is because we know we're in Arm mode. Something is removing the bottom bit of the fake PC value to reflect that. Which is probably fine.

The other failure is that we get no frames at all:
```
(Pdb) p thread.GetNumFrames()
0
(Pdb) p thread
python3: /home/david.spickett/llvm-project/lldb/source/Target/StackFrameList.cpp:687: StackFrameSP lldb_private::StackFrameList::GetFrameAtIndex(uint32_t): Assertion `!m_thread.IsValid() && "A valid thread has no frames."' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace and instructions to reproduce the bug.
#0 0xe1dc0758 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) Signals.cpp:0:0
#1 0xe1dbdc3c llvm::sys::RunSignalHandlers() Signals.cpp:0:0
#2 0xe1dc1784 SignalHandler(int, siginfo_t*, void*) Signals.cpp:0:0
#3 0xe84dfe50 __default_rt_sa_restorer ./signal/../sysdeps/unix/sysv/linux/arm/sigrestorer.S:80:0
#4 0xe84d06c6 __libc_do_syscall ./csu/../sysdeps/unix/sysv/linux/arm/libc-do-syscall.S:47:0
#5 0xe851069c __pthread_kill_implementation ./nptl/pthread_kill.c:44:76
#6 0xe84defc6 raise ./signal/../sysdeps/posix/raise.c:27:6
Fatal Python error: Aborted

Current thread 0xe8495020 (most recent call first):
  File "/home/david.spickett/build-llvm-arm/local/lib/python3.12/dist-packages/lldb/__init__.py", line 14960 in __repr__
  File "/usr/lib/python3.12/pdb.py", line 1335 in _msg_val_func
  File "/usr/lib/python3.12/pdb.py", line 1350 in do_p
  File "/usr/lib/python3.12/cmd.py", line 217 in onecmd
  File "/usr/lib/python3.12/pdb.py", line 527 in onecmd
  File "/usr/lib/python3.12/cmd.py", line 138 in cmdloop
  File "/usr/lib/python3.12/pdb.py", line 390 in _cmdloop
  File "/usr/lib/python3.12/pdb.py", line 425 in interaction
  File "/usr/lib/python3.12/pdb.py", line 329 in user_line
  File "/usr/lib/python3.12/bdb.py", line 114 in dispatch_line
  File "/usr/lib/python3.12/bdb.py", line 90 in trace_dispatch
  File "/home/david.spickett/llvm-project/lldb/test/API/functionalities/scripted_frame_provider/TestScriptedFrameProvider.py", line 42 in test_scripted_frame_objects
  File "/usr/lib/python3.12/unittest/case.py", line 589 in _callTestMethod
  File "/usr/lib/python3.12/unittest/case.py", line 634 in run
  File "/usr/lib/python3.12/unittest/case.py", line 690 in __call__
  File "/usr/lib/python3.12/unittest/suite.py", line 122 in run
  File "/usr/lib/python3.12/unittest/suite.py", line 84 in __call__
  File "/usr/lib/python3.12/unittest/suite.py", line 122 in run
  File "/usr/lib/python3.12/unittest/suite.py", line 84 in __call__
  File "/usr/lib/python3.12/unittest/runner.py", line 240 in run
  File "/home/david.spickett/llvm-project/lldb/packages/Python/lldbsuite/test/dotest.py", line 1164 in run_suite
  File "/home/david.spickett/llvm-project/lldb/test/API/dotest.py", line 8 in <module>

Extension modules: lldb._lldb (total: 1)
```
Though the provider was registered successfully.

`ScriptedFrameObjectProvider` is asked for frame 0, but nothing further.

Since you've got MacOS things to fix I won't dig into this now. Fix those and if it still fails on Arm I'll look at it again.
